fedimint-client/src/module/mod.rs

```diff
@@ -55,9 +68,174 @@ where
     }
+
+    fn input_amount(&self, input: &DynInput) -> TransactionItemAmount {
+        <T as ClientModule>::input_amount(
+            self,
+            input
+                .as_any()
+                .downcast_ref()
+                .expect("Dispatched to correct module"),
+        )
+    }
+
+    fn output_amount(&self, output: &DynOutput) -> TransactionItemAmount {
+        <T as ClientModule>::output_amount(
+            self,
+            output
+                .as_any()
+                .downcast_ref()
+                .expect("Dispatched to correct module"),
+        )
+    }
 }
 
 dyn_newtype_define!(
     #[derive(Clone)]
     pub DynClientModule(Arc<IClientModule>)
 );
+
+impl AsRef<maybe_add_send_sync!(dyn IClientModule + 'static)> for DynClientModule {
+    fn as_ref(&self) -> &maybe_add_send_sync!(dyn IClientModule + 'static) {
+        self.0.as_ref()
+    }
+}
+
+pub type StateGenerator<S> = Box<dyn Fn(TransactionId, u64) -> Vec<S>>;
+
+/// A client module that can be used as funding source and to generate arbitrary
+/// change outputs for unbalanced transactions.
+#[apply(async_trait_maybe_send!)]
+pub trait PrimaryClientModule: ClientModule {
+    // TODO: unclear if we should return a vec of inputs
+    /// Creates an input of **at least** a given `min_amount` from the holdings
+    /// managed by the module.
+    ///
+    /// If successful it returns:
+    /// * A set of private keys belonging to the input for signing the
+    ///   transaction
+    /// * The input of **at least** `min_amount`, the actual amount might be
+    ///   larger, the caller has to handle this case and possibly generate
+    ///   change using `create_change_output`.
+    /// * A closure that generates states belonging to the input. This closure
+    ///   takes the transaction id of the transaction in which the input was
+    ///   used and the input index as input since these cannot be known at time
+    ///   of calling `create_funding_input` and have to be injected later.
+    ///
+    /// The function returns an error if the client's funds are not sufficient
+    /// to create the requested input.
+    async fn create_funding_input(
+        &self,
+        dbtx: &mut DatabaseTransaction<'_>,
+        min_amount: Amount,
+    ) -> anyhow::Result<(
+        Vec<KeyPair>,
+        <Self::Common as ModuleCommon>::Input,
+        StateGenerator<Self::States>,
+    )>;
+
+    /// Creates an output of **exactly** `amount` that will pay into the
+    /// holdings managed by the module.
+    ///
+    /// It returns:
+    /// * The output of **exactly** `amount`.
+    /// * A closure that generates states belonging to the output. This closure
+    ///   takes the transaction id of the transaction in which the output was
+    ///   used and the output index as input since these cannot be known at time
+    ///   of calling `create_change_output` and have to be injected later.
+    async fn create_change_output(
```

Comment:
`create_exact_output`? Sure, we'll probably use it for creating change, but on a higher level we only care about module being able to create an exact arbitrary value output, and you even made that world bold already - why not just put it in a name then?